x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/SerializableTokenListCategory.java

```diff
@@ -160,6 +162,16 @@ public BytesRef[] getKeyTokens() {
         return Arrays.stream(keyTokenIndexes).mapToObj(index -> baseTokens[index]).toArray(BytesRef[]::new);
     }
 
+    public String getRegex() {
+        if (keyTokenIndexes.length == 0 || orderedCommonTokenBeginIndex == orderedCommonTokenEndIndex) {
+            return ".*";
```

Comment:
What does it mean when this condition is true? That the category contains no common terms? The category is an empty string?

My concern is for programmatic users of the API who use the regex `.*` without checking then have a very slow match of everything. What is the intended behaviour here

Reply:
In practice it means that there are some messages where the categorization analyzer kept 0 tokens. (In theory it could also mean that two categories with no common tokens got merged, but our validation on the similarity threshold should stop that - we don't allow similarity threshold 0 to be configured.) So the messages aren't necessarily empty strings, but consist entirely of characters that the categorization analyzer either didn't include in any token or discarded due to token filters.  Our default categorization analyzer would do this for a message that consisted of only numbers and whitespace, like `1 2 3 4` for example.  Then the categorization algorithm has no tokens to work with.

Setting the regex to `.*` in this situation is the same as what our C++ categorization has done for years - see https://github.com/elastic/ml-cpp/blob/2b0056ce742d3a6d9092b82e930664e1d5a65704/lib/model/CTokenListReverseSearchCreator.cc#L45.

The `max_matching_length` should protect us to some extent in this situation, as you'd hope that messages where the categorization analyzer finds no tokens should be rare and very short.  If the categorization analyzer is regularly finding no tokens then the chances are the data is inappropriate for categorization or else a different categorization analyzer needs to be used.

Reply:
The [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-categorize-text-aggregation.html) contain `max_matching_length` in the response objects but there is no description of what it means. Can you add a sentence to the response example pls describing `max_matching_length` and `regex`

Something like:
The `regex` field is a regular expression matching the category text (`key` field). Categories from short text messages containing few tokens may also match categories containing many tokens derived from longer text messages.  `max_matching_length` is the maximum length of the source messages in the category, all messages in the category are shorter than `max_matching_length`. Use this field to prevent the regular expression for small categories matching larger ones.